### PR TITLE
[New Feature] 클러스터링 기능 추가

### PR DESCRIPTION
**Pull request by note11g**
parent issue: #207 

바로 Merge하지 않고, 현재 f7fd068f663fd5e1b267db423b4a1979cdb520c6 까지의 작업을 토대로 인터페이스 검토 후 구현을 진행할 예정이며, PR은 구현 이후 Merge합니다.

### 클러스터블 마커 추가
클러스터링 가능한 마커의 추가. 마커와 동일하게 사용합니다.
```dart
final restaurants = await RestaurantAPI.requestBy(region: "Seoul");
final markers = restaurants.map(
    (restaurant) => NClusterableMarker(
      id: restaurant.id,
      position: restaurant.position,
      icon: restaurantIcon,
      caption: NOverlayCaption(text: restaurant.name),
      tags: { // type: Map<String, String>
        "gu": restaurant.addr.gu, // 추후 구/동 별로 클러스터링 병합 전략을 지정할 수 있습니다.
        "dong": restaurant.addr.dong,
      },
  )..setOnTapListener((NClusterableMarker cm) => showRestaurantInfo(cm.info)));

await mapController.addOverlayAll(markers.toSet());
```

### 고급 사용법
```dart
NaverMap(
  clusterOptions: NaverMapClusterOptions(
    animate: true,
    enableZoomLevel: const NRange(null, 15, includeMax: false), // 15 미만으로 줌 아웃 시 클러스터링
    clusterMarkerBuilder: (info, clusterMarker) {
      final (size, color) = switch (info.size) {
        >= 100 => (56.0, Colors.redAccent),
        >= 35 => (52.0,  Colors.orange),
        _ => (48.0,  Colors.green),
      };
      clusterMarker
        ..setIcon(circleIcon)
        ..setTintColor(color)
        ..setSize(NSize(size, size))
        ..setCaption(info.mergedTag != null ? NOverlayCaption(
            text: info.mergedTag,
            color: Colors.white,
            haloColor:color
            textSize: 14,
            requestWidth: size,
          ) : null)
        ..setSubCaption(NOverlayCaption(
            text: info.size.toString(), 
            color: Colors.white,
            haloColor: Colors.transparent,
         ));
    },
    mergeStrategy: const NClusterMergeStrategy(
      // (4개 미만이면 클러스터링 하지 않음) 클러스터링 대상으로 병합할 최소 마커 수. default: 2
      minClusterChildCount: 4,
      // 클러스터링 대상으로 계산할 최소 화면 거리. default: 70dp
      willMergedScreenDistance: 100,
      // 해당 화면 거리 이상의 태그는 클러스터링 대상으로 계산하지 않습니다. default: ((min(screenW, screenH) * 2 / 3)dp
      maxMergedScreenDistanceByTag: 200,
      // 태그별로 병합이 동작할 줌 레벨을 지정. 화면 거리 기반 병합보다 우선됨. type: Map<String, NRange>
      mergeByEachTagEnableZoomLevel: {
        "gu": NRange(10, 12, includeMax: false), // 구별로 클러스터링
        "dong": NRange(12, 14), // 동별로 클러스터링
      },
    ),
  ),
);
```

### CODE REVIEW

File: lib/src/widget/map_widget.dart [hunk 1]
```diff
@@ -61,6 +66,7 @@ class NaverMap extends StatefulWidget {
   const NaverMap({
     super.key,
     this.options = const NaverMapViewOptions(),
+    this.clusterOptions = const NaverMapClusterOptions(),
```
Comment (mrkim0425) [inline lib/src/widget/map_widget.dart:69]: line 12 : 기본 값은 클러스터링 옵션을 사용하지 않는 것을 의미하는 null입니다.
Based on these sentence line 69's constructor default value should be null but its' not.

File: lib/src/type/map/overlay/overlay/addable/marker_wrapper.dart
Comment (mrkim0425) [inline lib/src/type/map/overlay/overlay/addable/marker_wrapper.dart:1]: Could you explain in more details why this wrapper class is necessary and what role is it for

File: lib/src/widget/map_widget.dart (hunk 1 above)
Comment (note11g) [reply lib/src/widget/map_widget.dart:69]: ok it's my mistake. i will fix it.

Comment (note11g) [reply lib/src/widget/map_widget.dart:69]: *non-nullable is right(my intent).

Q) why it's non-null? 
A) because clustering only turned on `NClusterableMarker` is added on map. that's why `null` value is not necessary on this api property.

File: lib/src/type/map/overlay/overlay/addable/marker_wrapper.dart
Comment (note11g) [reply lib/src/type/map/overlay/overlay/addable/marker_wrapper.dart:1]: Here's a response explaining the purpose of the `_NMarkerWrapper` class:

The `_NMarkerWrapper` class is a wrapper around the `NMarker` class, serving two primary purposes:

1. **Extension over Inheritance**: Instead of directly modifying or extending the `NMarker` class, the `_NMarkerWrapper` inherits from the `NAddableOverlay` superclass and implements the `NMarker` interface. This approach allows for extending the functionality of `NMarker` without modifying its codebase directly, promoting better code organization and maintainability. (To be honest, when designing the overlay structure two years ago, I did not design it in a way that allowed for potential subclasses of overlay implementations like the marker. At that time, I used an enum for type checking, but this approach was not suitable for explicitly defining the class hierarchy.)

2. **API Compatibility and Preservation**: While it would have been possible to use generics like `NMarker<SubType extends NMarker> extends NAddableOverlay<SubType>` for a cleaner implementation, this approach could potentially decrease type usability (changing the base type from `NMarker` to `NMarker<NMarker<void>>`). To maintain API compatibility and avoid breaking existing code, the `_NMarkerWrapper` overrides the `setOnTapListener` method, ensuring that the existing API remains functional.

However, as mentioned, there is an ongoing consideration regarding refactoring the constructor specification of `NAddableOverlay` in a way that does not compromise compatibility. This potential refactoring aims to improve the code structure while preserving backwards compatibility.

In summary, the `_NMarkerWrapper` class acts as an intermediary layer, allowing for the extension of `NMarker` functionality without directly modifying its codebase, while also maintaining API compatibility and ensuring existing code continues to work as expected.

File: lib/src/widget/map_widget.dart (hunk 1 above)
Comment (note11g) [reply lib/src/widget/map_widget.dart:71]: fixed on da02e991dfe5b6fec70ce881372a34dac18cc4a9

File: lib/src/type/map/overlay/overlay/addable/marker_wrapper.dart
Comment (mrkim0425) [reply lib/src/type/map/overlay/overlay/addable/marker_wrapper.dart:1]: Thank you for the explanation :)
LGTM, Thanks for your effort

**Review by mrkim0425 — APPROVED** [review]

**Comment by note11g** [comment]
구현을 진행합니다.

**Comment by note11g** [comment]
클러스터링 기능의 Preview 버전의 출시일을 안내드립니다.
Android Only Preview: 검수 이전 버전 5/6(월), 검수 이후 버전 5/7(화)
All Platform(Android+iOS) Support Preview: 5/8(수)

이후 사용자 피드백을 수렴하여 5월 3주차에 정식으로 1.3.0 버전이 출시될 예정입니다.

**Comment by note11g** [comment]
검수 과정에서 설계상 이슈가 발견되어, 출시가 조금 더 늦어질 것으로 예상됩니다.
정식출시는 미루어지지 않으며, Preview 버전만 이번주 내로 출시되는 것으로 일정이 미루어집니다.
기다려주시는 많은 분들께 죄송하다는 말씀드립니다.

**Comment by note11g** [comment]
테스트코드는 최종 병합 이전 단계에서 작성하겠습니다.

- 현재 known issues
iOS Memory Leak by multiple `addOverlay(clusterableMarker)` called. (not with `overlayAll(clusterableMarker)` called once)
